### PR TITLE
Support position_ids as input for flash attention

### PR DESCRIPTION
**Pull request by shadow150519**

### CODE REVIEW

File: torch_xla/csrc/flash_attention_utils.cpp [hunk 1]
```diff
@@ -382,9 +389,16 @@ FlashAttentionBackwardParams get_flash_attention_backward_params(
     TORCH_CHECK(cu_seqlens_k.value().dtype() == torch::kInt32);
     TORCH_CHECK(cu_seqlens_q.value().is_contiguous());
     TORCH_CHECK(cu_seqlens_k.value().is_contiguous());
-    TORCH_CHECK(batch_size == cu_seqlens_q.value().numel() - 1);
-    CHECK_SHAPE(cu_seqlens_q.value(), batch_size + 1);
-    CHECK_SHAPE(cu_seqlens_k.value(), batch_size + 1);
+    TORCH_CHECK(batch_size == cu_seqlens_q.value().numel() - 1 ||
+                batch_size == 1);
```
Comment (Seventeen17) [inline torch_xla/csrc/flash_attention_utils.cpp]: Add a comment or TODO that if batch_size > 1 is supported in the future, need to be modified here.

File: torch_xla/csrc/flash_attention_utils.cpp [hunk 2]
```diff
@@ -382,9 +389,16 @@ FlashAttentionBackwardParams get_flash_attention_backward_params(
...
+    TORCH_CHECK(
+        cu_seqlens_q.value().sizes() == torch::IntArrayRef({batch_size + 1}) ||
+            cu_seqlens_q.value().sizes() == torch::IntArrayRef({seqlen_q + 1}),
```
Comment (Seventeen17) [inline torch_xla/csrc/flash_attention_utils.cpp]: Maybe it is seqlen_q * batch_size + 1?

File: torch_xla/csrc/flash_attention_utils.cpp [hunk 3]
```diff
@@ -421,11 +450,12 @@ at::Tensor cu_seqlens_to_indices(const at::Tensor& cu_seqlens, int batch_size,
...
-      torch::arange(shape[0], opts.dtype(torch::kInt32)).unsqueeze(1);
-  torch::Tensor cols =
-      torch::arange(shape[1], opts.dtype(torch::kInt32)).unsqueeze(0);
-  torch::Tensor mask = cols < nonzero_counts.unsqueeze(1);
+  torch::Tensor rows = torch::arange(shape[0], opts.dtype(torch::kInt32))
+                           .unsqueeze(1);  // (batch_size,1)
+  torch::Tensor cols = torch::arange(shape[1], opts.dtype(torch::kInt32))
+                           .unsqueeze(0);  // (1,seqlen)
+  torch::Tensor mask =
+      cols < nonzero_counts.unsqueeze(1);  // (1,seqlen) < (batch_size, 1)
```
Comment (Seventeen17) [inline torch_xla/csrc/flash_attention_utils.cpp]: Remove the comments or rewrite them to be more general.

File: torch_xla/csrc/flash_attention_utils.cpp [hunk 4]
```diff
@@ -451,18 +481,86 @@ at::Tensor mask_to_indices(const at::Tensor& attention_mask,
...
+  int total, max_seqlen_in_batch;
+  at::Tensor valid_cu_seqlens = cu_seqlens.slice(0, 0, batch_size + 1);
+  at::Tensor indices =
+      cu_seqlens_to_indices(valid_cu_seqlens, batch_size, max_seqlen,
+                            torch::kInt64, max_seqlen_in_batch, total);
+  at::Tensor result = at::empty({total, nhead}, pad_softmax_lse.options());
+  result.copy_(pad_softmax_lse.transpose(1, 2)
+                   .reshape({batch_size * max_seqlen, nhead})
+                   .index({indices, torch::indexing::Slice()}));
+  return result.transpose(0, 1).unsqueeze(0);
```
Comment (Seventeen17) [inline torch_xla/csrc/flash_attention_utils.cpp:505]: Add a comment that if batch_size > 1 is supported in the future, need to be modified here.

File: torch_xla/csrc/flash_attention_utils.cpp [hunk 5]
```diff
@@ -3,6 +3,8 @@
 #include <ATen/cuda/CUDAContext.h>
 #include <torch/extension.h>
 
+#include <iostream>
```
Comment (Seventeen17) [inline torch_xla/csrc/flash_attention_utils.cpp]: Remove if not needed.

File: torch_xla/csrc/flash_attention_utils.cpp [hunk 6]
```diff
@@ -451,18 +481,86 @@ at::Tensor mask_to_indices(const at::Tensor& attention_mask,
...
+  int total;
+  at::Tensor valid_cu_seqlens = cu_seqlens.slice(0, 0, batch_size + 1);
+  at::Tensor indices =
+      cu_seqlens_to_indices(valid_cu_seqlens, batch_size, max_seq_len,
+                            torch::kInt32, max_seqlen_in_batch, total);
+  TORCH_CHECK(indices.size(0) == softmax_lse.size(2),
+              "indice should be same size with softmax_lse")
+
+  at::Tensor result =
+      at::empty({batch_size * max_seq_len, nheads}, softmax_lse.options());
```
Comment (Seventeen17) [inline torch_xla/csrc/flash_attention_utils.cpp]: use `at::zeros` to init with zeros?

File: torch_xla/csrc/ops/flash_attention_varlen_position_ids_forward.cpp [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+                       opts.dtype(scalar_type));
+  at::Tensor cu_seqlens_q =
+      torch::from_blob(buffers[7 + buf_offset], {params.seqlen_q + 1}, opts);
+  at::Tensor cu_seqlens_k =
+      torch::from_blob(buffers[8 + buf_offset], {params.seqlen_k + 1}, opts);
+  at::Tensor rng_state =
+      torch::from_blob(buffers[6 + buf_offset], {2}, opts.dtype(torch::kInt64));
+  softmax_lse.fill_(0);
+  o_output.fill_(0);
+  cu_seqlens_k.fill_(-1);
```
Comment (Seventeen17) [inline torch_xla/csrc/ops/flash_attention_varlen_position_ids_forward.cpp]: Move the `fill_(-1)` in function `position_ids_to_indices` incase the initialization missed outside the function.

File: torch_xla/csrc/ops/flash_attention_varlen_position_ids_forward.cpp [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+  int total_k = params.b * params.seqlen_k;
+  at::Tensor indices_k;
+
+  int real_batch_size;  // packed qkv's batch size is 1, but fa need to know
+                        // real batch size before packing.
+  indices_k = position_ids_to_indices(position_ids, max_seqlen_in_batch_k,
+                                      total_k, cu_seqlens_k, real_batch_size);
+  TORCH_CHECK(cu_seqlens_k.size(0) == params.seqlen_k + 1,
+              "cu_seqlen'shape should be params.seqlen_k");
+  TORCH_CHECK(indices_k.dtype() == torch::kInt64, "indice should be int64");
```
Comment (Seventeen17) [inline torch_xla/csrc/ops/flash_attention_varlen_position_ids_forward.cpp:108]: Is kInt32 satisfied?

File: torch_xla/csrc/ops/flash_attention_varlen_forward.cpp [hunk 9]
```diff
@@ -4,6 +4,8 @@
 #include <c10/cuda/CUDAGuard.h>
 #include <torch/extension.h>
 
+#include <iostream>
+
```
Comment (Seventeen17) [inline torch_xla/csrc/ops/flash_attention_varlen_forward.cpp]: Remove if not needed

File: torch_xla/csrc/ops/flash_attention_varlen_position_ids_backward.cpp [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+  at::Tensor k = torch::from_blob(
+      buffers[2], {params.b * params.seqlen_k, params.h_k, params.d}, opts);
+  at::Tensor v = torch::from_blob(
+      buffers[3], {params.b * params.seqlen_k, params.h_k, params.d}, opts);
+  at::Tensor o = torch::from_blob(
+      buffers[4], {params.b * params.seqlen_q, params.h, params.d}, opts);
+  at::Tensor softmax_lse =
+      torch::from_blob(buffers[5], {params.b, params.h, params.seqlen_q},
+                       opts.dtype(torch::kFloat));
+  at::Tensor cu_seqlens_q = torch::from_blob(buffers[6], {params.seqlen_q + 1},
```
Comment (Seventeen17) [inline torch_xla/csrc/ops/flash_attention_varlen_position_ids_backward.cpp]: Init with size `params.b * params.seqlen + 1` incase batch_size >1 in the future?

File: torch_xla/csrc/ops/flash_attention_varlen_position_ids_forward.cpp [hunk 11]
```diff
@@ -0,0 +1,331 @@
+#include "torch_xla/csrc/ops/flash_attention_varlen_position_ids_forward.h"
+
+#include <ATen/cuda/CUDAContext.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <torch/extension.h>
+
+#include <iostream>
```
Comment (Seventeen17) [inline torch_xla/csrc/ops/flash_attention_varlen_position_ids_forward.cpp]: Remove?

**Comment by Seventeen17** [comment]
👍

File: torch_xla/csrc/flash_attention_utils.cpp (hunk 3 above)
Comment (shadow150519) [reply torch_xla/csrc/flash_attention_utils.cpp]: done

File: torch_xla/csrc/flash_attention_utils.cpp (hunk 1 above)
Comment (shadow150519) [reply torch_xla/csrc/flash_attention_utils.cpp]: done

File: torch_xla/csrc/flash_attention_utils.cpp (hunk 2 above)
Comment (shadow150519) [reply torch_xla/csrc/flash_attention_utils.cpp]: done

File: torch_xla/csrc/flash_attention_utils.cpp (hunk 6 above)
Comment (shadow150519) [reply torch_xla/csrc/flash_attention_utils.cpp]: done

File: torch_xla/csrc/flash_attention_utils.cpp (hunk 5 above)
Comment (shadow150519) [reply torch_xla/csrc/flash_attention_utils.cpp]: done

File: torch_xla/csrc/flash_attention_utils.cpp (hunk 4 above)
Comment (shadow150519) [reply torch_xla/csrc/flash_attention_utils.cpp:505]: done

File: torch_xla/csrc/ops/flash_attention_varlen_position_ids_forward.cpp (hunk 7 above)
Comment (shadow150519) [reply torch_xla/csrc/ops/flash_attention_varlen_position_ids_forward.cpp]: done

File: torch_xla/csrc/ops/flash_attention_varlen_position_ids_forward.cpp (hunk 8 above)
Comment (shadow150519) [reply torch_xla/csrc/ops/flash_attention_varlen_position_ids_forward.cpp:108]: pytorch need index tensor to be torch::kInt64

File: torch_xla/csrc/ops/flash_attention_varlen_forward.cpp (hunk 9 above)
Comment (shadow150519) [reply torch_xla/csrc/ops/flash_attention_varlen_forward.cpp]: done

File: torch_xla/csrc/ops/flash_attention_varlen_position_ids_backward.cpp (hunk 10 above)
Comment (shadow150519) [reply torch_xla/csrc/ops/flash_attention_varlen_position_ids_backward.cpp]: done

File: torch_xla/csrc/ops/flash_attention_varlen_position_ids_forward.cpp (hunk 11 above)
Comment (shadow150519) [reply torch_xla/csrc/ops/flash_attention_varlen_position_ids_forward.cpp]: done

**Review by Seventeen17 — APPROVED** [review]

**Pull request merged**